module/state_synchronization/indexer/persister.go

```diff
@@ -0,0 +1,221 @@
+	log = log.With().
+		Str("component", "persister").
+		Hex("block_id", logging.ID(executionResult.BlockID)).
+		Uint64("height", header.Height).
+		Logger()
+
+	persister := &Persister{
+		log:                    log,
+		inMemoryRegisters:      inMemoryRegisters,
+		inMemoryEvents:         inMemoryEvents,
+		inMemoryCollections:    inMemoryCollections,
+		inMemoryTransactions:   inMemoryTransactions,
+		inMemoryResults:        inMemoryResults,
+		inMemoryTxResultErrMsg: inMemoryTxResultErrMsg,
+		registers:              registers,
+		events:                 events,
+		collections:            collections,
+		transactions:           transactions,
+		results:                results,
+		txResultErrMsg:         txResultErrMsg,
+		executionResult:        executionResult,
+		header:                 header,
+		protocolDB:             protocolDB,
+	}
+
+	persister.log.Info().
+		Uint64("height", header.Height).
+		Msg("persister initialized")
+
+	return persister
+}
+
+// Persist save data from in-memory storages to the provided persisted storages and commit updates to the database.
+// No errors are expected during normal operations
+func (p *Persister) Persist() (err error) {
+	// Create a batch for atomic updates
+	batch := p.protocolDB.NewBatch()
+
+	defer func() {
+		err = multierr.Combine(err, batch.Close())
+	}()
+
+	p.log.Debug().Msg("adding execution data to batch")
+
+	start := time.Now()
+
+	if err = p.persistRegisters(); err != nil {
+		return
+	}
+
+	if err = p.addEventsToBatch(batch); err != nil {
+		return
+	}
+
+	if err = p.addResultsToBatch(batch); err != nil {
+		return
+	}
+
+	if err = p.addCollectionsToBatch(batch); err != nil {
+		return
+	}
+
+	if err = p.addTransactionsToBatch(batch); err != nil {
+		return
+	}
+
+	if err = p.addTransactionResultErrorMessagesToBatch(batch); err != nil {
+		return
+	}
+
+	// TODO: include update to latestPersistedSealedResultBlockHeight in the batch
+
+	if err = batch.Commit(); err != nil {
+		err = fmt.Errorf("failed to commit batch: %w", err)
+		return
+	}
+
+	duration := time.Since(start)
+
+	p.log.Debug().
+		Dur("duration_ms", duration).
+		Int("events_count", len(p.inMemoryEvents.Data())).
+		Int("collection_count", len(p.inMemoryCollections.Data())).
+		Int("transactions_count", len(p.inMemoryTransactions.Data())).
+		Int("transaction_results_count", len(p.inMemoryResults.Data())).
+		Int("transaction_result_error_messages_count", len(p.inMemoryTxResultErrMsg.Data())).
+		Msg("successfully prepared execution data for persistence")
+
+	return
+}
+
+// persistRegisters persists registers from in-memory to permanent storage.
+// Registers must be stored for every height, even if it's an empty set.
+// No errors are expected during normal operations
+func (p *Persister) persistRegisters() error {
+	registerData := p.inMemoryRegisters.Data()
+
+	// Always store registers for every height to maintain height continuity
+	if err := p.registers.Store(registerData, p.header.Height); err != nil {
```

Comment:
The registerData and header height are stored separately, not sure if there is any synchronization. Is it possible the registerData include the height itself.

Can we add some sanity check that p.header.Height matches with registerData.Height?

Or we could do:
```
	registerData := p.inMemoryRegisters.Data(p.header.Height)
```
And let the inMemoryRegisters to ensure the height is correct.